### PR TITLE
Fixes #28506 - fix upgraded prometheus

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Someone forgot to check the changelog before theforeman/foreman-packaging#4439 :eyes:

### CODE REVIEW

File: lib/foreman/telemetry_sinks/prometheus_sink.rb [hunk 1]
```diff
@@ -6,15 +6,15 @@ def initialize(opts = {})
     end
 
     def add_counter(name, description, instance_labels)
-      @prom.counter(name.to_sym, description)
+      @prom.counter(name.to_sym, docstring: description)
```
Comment (ezr-ondrej) [inline lib/foreman/telemetry_sinks/prometheus_sink.rb]: @lzap should we pass the instance_labels? Prometheus client seems to have something called [pretty similarily](https://github.com/prometheus/client_ruby/blob/v1.0.0/lib/prometheus/client/registry.rb), but I have no clue if there is a reason we don't pass them on?

**Comment by ekohl** [comment]
Thanks for figuring this out. Was on the list of things I wanted to know. Given this was found in puppet-foreman by actually deploying, it appears there are no tests to cover this.

**Comment by ezr-ondrej** [comment]
Yeah it appears so, it would be good to add them, to know it sooner, but
I'm mostly AFK this week, so I'll not be able to write them.

Dne so 14. 12. 2019 14:02 uživatel Ewoud Kohl van Wijngaarden <
notifications@github.com> napsal:

**Review by lzap — CHANGES_REQUESTED** [review]
Yeah thanks. Please incorporate the following:

```diff
diff --git a/config/initializers/5_telemetry.rb b/config/initializers/5_telemetry.rb
index 22570a6c0..eb2f4cc11 100644
--- a/config/initializers/5_telemetry.rb
+++ b/config/initializers/5_telemetry.rb
@@ -1,3 +1,11 @@
+# Set multiprocess-friendly data store
+require 'prometheus/client'
+require 'prometheus/client/data_stores/direct_file_store'
+PROMETHEUS_STORE_DIR = File.join(Rails.root, 'tmp', 'prometheus')
+Dir.mkdir(PROMETHEUS_STORE_DIR) unless Dir.exist?(PROMETHEUS_STORE_DIR)
+Prometheus::Client.config.data_store =
+  Prometheus::Client::DataStores::DirectFileStore.new(dir: PROMETHEUS_STORE_DIR)
+
 # Foreman telemetry global setup.
 telemetry = Foreman::Telemetry.instance
 if SETTINGS[:telemetry] && (Rails.env.production? || Rails.env.development?)
diff --git a/config/initializers/5_telemetry_metrics.rb b/config/initializers/5_telemetry_metrics.rb
index c4c4d0c35..2d5f56361 100644
--- a/config/initializers/5_telemetry_metrics.rb
+++ b/config/initializers/5_telemetry_metrics.rb
@@ -7,7 +7,7 @@
 # Plugins can add own metrics through add_counter_telemetry, add_gauge_telemetry and add_histogram_telemetry.
 #
 telemetry = Foreman::Telemetry.instance
-telemetry.add_counter(:http_requests, 'A counter of HTTP requests made', [:controller, :action])
+telemetry.add_counter(:http_requests, 'A counter of HTTP requests made', [:controller, :action, :status])
 telemetry.add_histogram(:http_request_total_duration, 'Total duration of controller action', [:controller, :action])
 telemetry.add_histogram(:http_request_db_duration, 'Time spent in database for a request', [:controller, :action])
 telemetry.add_histogram(:http_request_view_duration, 'Time spent in view for a request', [:controller, :action])
diff --git a/lib/foreman/telemetry.rb b/lib/foreman/telemetry.rb
index 7efa9b337..2d80d50c8 100644
--- a/lib/foreman/telemetry.rb
+++ b/lib/foreman/telemetry.rb
@@ -7,7 +7,7 @@ module Foreman
     extend Forwardable
     attr_accessor :prefix, :sinks
 
-    DEFAULT_BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10].freeze
+    DEFAULT_BUCKETS = [1, 5, 20, 50, 100, 250, 500, 2000].freeze
 
     def initialize
       @sinks = []
diff --git a/lib/foreman/telemetry_sinks/prometheus_sink.rb b/lib/foreman/telemetry_sinks/prometheus_sink.rb
index 2d804d31c..855957921 100644
--- a/lib/foreman/telemetry_sinks/prometheus_sink.rb
+++ b/lib/foreman/telemetry_sinks/prometheus_sink.rb
@@ -6,27 +6,27 @@ module Foreman::TelemetrySinks
     end
 
     def add_counter(name, description, instance_labels)
-      @prom.counter(name.to_sym, docstring: description)
+      @prom.counter(name.to_sym, docstring: description, labels: instance_labels)
     end
 
     def add_gauge(name, description, instance_labels)
-      @prom.gauge(name.to_sym, docstring: description)
+      @prom.gauge(name.to_sym, docstring: description, labels: instance_labels)
     end
 
     def add_histogram(name, description, instance_labels, buckets)
-      @prom.histogram(name.to_sym, docstring: description, buckets: buckets)
+      @prom.histogram(name.to_sym, docstring: description, buckets: buckets, labels: instance_labels)
     end
 
     def increment_counter(name, value, tags)
-      @prom.get(name.to_sym).increment(tags, value)
+      @prom.get(name.to_sym).increment(by: value, labels: tags)
     end
 
     def set_gauge(name, value, tags)
-      @prom.get(name.to_sym).set(tags, value)
+      @prom.get(name.to_sym).set(value, labels: tags)
     end
 
     def observe_histogram(name, value, tags)
-      @prom.get(name.to_sym).observe(tags, value)
+      @prom.get(name.to_sym).observe(value, labels: tags)
     end
   end
 end
```

**Comment by ekohl** [comment]
@lzap mind submitting a PR with that so we can merge it? @ezr-ondrej said he's AFK most of the week.

**Comment by ezr-ondrej** [comment]
@lzap I've done it, but I don't feel really comfortable about it as I know very little about prometheus :eyes: if you would like to take over and make it better, it would be appreciated.

File: config/initializers/5_telemetry.rb [hunk 2]
```diff
@@ -1,3 +1,11 @@
+# Set multiprocess-friendly data store
+require 'prometheus/client'
+require 'prometheus/client/data_stores/direct_file_store'
+PROMETHEUS_STORE_DIR = File.join(Rails.root, 'tmp', 'prometheus')
+Dir.mkdir(PROMETHEUS_STORE_DIR) unless Dir.exist?(PROMETHEUS_STORE_DIR)
```
Comment (ekohl) [inline config/initializers/5_telemetry.rb]: This causes a failure:
```
Errno::ENOENT: No such file or directory @ dir_s_mkdir - /home/jenkins/workspace/test_develop_pr_core@2/database/postgresql/ruby/2.6/slave/fast/tmp/prometheus
```

```suggestion
FileUtils.mkdir_p(PROMETHEUS_STORE_DIR)
```

Or do it manually like in the secret token handling:
https://github.com/theforeman/foreman/blob/739ed809538dcc3919d6614aba44fc90b59a7ef5/config/initializers/secret_token.rb#L5-L6

**Review by ekohl — DISMISSED** [review]

**Comment by lzap** [comment]
Well I would but as you have found out it currently does not work and you can't make it worse! :-) So if you have time, please incorporate Ewoud comment and let's fix this.

I've included one important change - it's now multi-process friendly. That was actually the feature we've been waiting for.

**Comment by ekohl** [comment]
Added a commit that uses `mkdir_p` to test it in CI. Should be squashed on merge.

**Comment by lzap** [comment]
Katello failure irrelevant.

**Review by lzap — APPROVED** [review]

File: config/initializers/5_telemetry.rb [hunk 3]
```diff
@@ -1,3 +1,11 @@
+# Set multiprocess-friendly data store
+require 'prometheus/client'
```
Comment (ekohl) [inline config/initializers/5_telemetry.rb:2]: Doesn't this break all installations where the telemetry group is unavailable?

Comment (lzap) [reply config/initializers/5_telemetry.rb:2]: https://github.com/theforeman/foreman/pull/7271

Comment (lzap) [reply config/initializers/5_telemetry.rb:2]: https://github.com/theforeman/foreman/pull/7271

**Comment by tbrisker** [comment]
1.24 - 49dd8328a 
please open a PR to update packaging in rpm/stable.

**Pull request merged**